mut/redirects/redirect_main.py

```diff
@@ -0,0 +1,396 @@
+            line = line.strip()
+            
+            # regex to see if we are dealing with a keyword - define, symlink, or raw
+            if re.search('(define|symlink|raw)', line):
+                keyword_split = line.split(':', 1)
+                
+                # define:
+                if keyword_split[0] == 'define':
+                    type_split = keyword_split[1].split(' ')
+                    
+                    key = type_split[1]
+                    
+                    
+                    if len(type_split) > 3:
+                        
+                        value = []
+                        
+                        for x in range (2, len(type_split)):
+                            value.append(type_split[x])
+                            
+                    else:
+                        value = type_split[2]
+                        
+                    definition = [key, value]
+                    
+                    global definitions
+                    definitions.append(definition)
+                    
+                    # see if any definitions have to be substituted with other definitions:
+                    for x in range(0, len(definitions)):
+                        sub_regex = '{(.*?)}'
+                        
+                        if type(definitions[x][1]) is str:
+                            matches = re.findall(sub_regex, definitions[x][1], re.DOTALL)
+    
+                            if matches:
+                                # loop through each match
+                                for y in range(0, len(matches)): 
+                                    # loop through each definition
+                                    for m in range(0, len(definitions)):
+                                        # if the match == the definition key
+                                        if matches[y] == definitions[m][0]:
+                                            # substitute the definition value for the match
+                                            if type(definitions[m][1]) is str:
+                                                definitions[x][1] = definitions[x][1].replace('${' + matches[y] + '}', definitions[m][1])
+                        
+                    
+                    versions = parse_versions(definitions)
+                    
+                # grab symlinks:
+                if keyword_split[0] == 'symlink':
+                    type_split = line.split(':', 1)
+                    sym_split = type_split[1].split('->')
+                    
+                    for x in range(0, len(sym_split)):
+                        sym_split[x] = sym_split[x].strip()
+                        
+                    global symlinks
+                    symlinks.append(sym_split)
+    
+                # raw redirects:
+                if keyword_split[0] == 'raw':
+                    p = re.compile('(?:[ \t\f\v])(.*)(?:[ \t\f\v]->)(.*)')
```

Comment:
Compile regular expressions once in the global scope if possible, rather than compiling in a loop like this.